### DISCUSSION TITLE
attach custom actions to TIMING_CLOCK

### DISCUSSION THREAD

**Discussion by sallen881** [tttapa/Control-Surface#637] [Q&A]
Hello! Thanks for all of your hard work on this excellent library! I've successfully built a bluetooth mixer that's a lot of fun to connect to software (VCV Rack and TouchDesigner in my case). Now I'm developing a second project, a sequencer. I have a downscaled prototype (four buttons) with very basic functionality working as intended, but my code is timed by the microcontroller's internal clock.

Is there a way to restructure the "Custom-note-LED-Input-Element-Callback" example to trigger actions on receiving a TIMING_CLOCK signal? The `TwoByteMIDIMatcher` in lines 45-46 is not appropriate in this case - timing clock data actually has zero bytes! My intuition was to change from a `MatchingMIDIInputElement` which takes Type and Matcher to just a `MIDIInputElement` which only takes Type:

`: public MatchingMIDIInputElement<MIDIMessageType::NOTE_ON,TwoByteMIDIMatcher> {`
becomes
`  : public MIDIInputElement<MIDIMessageType::TIMING_CLOCK> {
`

However I have been struggling to get this change compiled. I am pretty sure it does not work because `MIDIInputElement` does not have a handleUpdate() function like `MatchingMIDIInputElement` does. However, I'm not sure where to make changes to fix this, or if I am going about it wrong. The code is rather advanced for me, I'm new to the concepts of class and constructor. Any tips would be appreciated!

-s

**Comment by tttapa** [comment]
It's probably easiest to use [this example](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d32/MIDI-Input-Callback_8ino-example.html) or [this example](https://tttapa.github.io/Control-Surface-doc/Doxygen/da/d27/MIDI-Input-Fine-Grained-All-Callbacks_8ino-example.html) if you just need MIDI. 

Once you have a `RealTimeMessage` you can use `rt.getMessageType() == MIDIMessageType::TIMING_CLOCK` to check whether it's a clock message.
